[CDAP-9415] Switch from font icons to SVG

tonybach opened the pull request:
JIRA: https://issues.cask.co/browse/CDAP-9415. 

### General
This PR is the first step in moving from using font icons to SVG. So far I have converted most of the fast actions and custom icons in the home list view and in detail views. I have also converted icons in the Cask Market and in the Plus Button modal.

I used IcoMoon to convert our existing font icons to SVGs, then put them in a file called `symbol-defs.svg`. I then inject the contents of the file into the HTML. To reduce boilerplate code, I have created an `IconSVG` component. 

I have also added svg minifying tasks to the Gulpfile, and added a polyfill for browsers that don't support SVG images.

### JSX Changes
Changed our existing code:

``` <span/i className="fa fa-{icon}" />```

to

``` <IconSVG value='icon-{icon}' />```

For floating icon spans (e.g. `<span className="fa fa-{icon} float-xs-right" />`), have to keep the `<span className="float-xs-right">`, then put the `IconSVG value="icon-{icon}"` inside.

However, some icons had to be changed to show correctly (e.g. `fa-times` -> `icon-close`, `fa-scissors` -> `icon-cut`). Not exactly sure why this is yet. Also for special `fa` classes like `fa-spin` or `fa-lg`, I'm still passing them as classes to the SVG e.g. `SVG className="fa-spin"`. Not sure if this is the right approach.

### HTML/CSS Changes
As expected, there were several CSS changes needed to ensure the same look as before. Here are the main ones I have discovered:
- `fa-{icon}` classes must be changed to `icon-{icon}`.
- `padding` of icons must be changed to `margin`.
- `font-size` must be changed to `height` & `width`.

### Todo:

- [x] Replace more font icons with SVGs.

- [x] Figure out how to reduce the number of network calls to the `/cdap_assets/fonts/symbol-defs.svg` file.

tonybach committed:
[CDAP-9415] Switch from font icons to SVG icons, in Home page and Market
- Minifies SVGs as a Gulp task
- Injects svg symbol definitions to the DOM, instead of making an external request for the file
- Adds plugins to allow es6 spread operator

elfenheart commented on cdap-ui/app/cdap/components/SVG/index.js:
```diff
@@ -0,0 +1,39 @@
+  return (
+    <svg
+      className={classnames('icon-svg', icon, className)}
+      onClick={onClick}
```
perhaps this should just be a spread operator that will take the rest of the props passed? Then it can just follow other all the other react props?

tonybach replied:
Not exactly sure how I can do this. Do you propose that I do this:
``` 
<svg
    className={classnames('icon-svg', icon, className)}
    ...this.props
>
```

tonybach replied:
Nvm, found a good article explaining how to do that here: http://blog.goguardian.com/nerds/an-odyssey-to-find-a-sustainable-icon-system-with-svgs-in-react

ajainarayanan commented on cdap-ui/app/cdap/components/EntityCard/EntityCardHeader/index.js:
```diff
@@ -55,7 +54,10 @@ export default class EntityCardHeader extends Component {
               >
                 <h4>
-                  <span className={classNames('entity-icon', this.props.entity.icon)}></span>
+                  <SVG
```
It would be good to call it something else, like `<FontContainer .../>` since `svg` is already an element. Even though HTML is case insensitive it would be good if we follow react/html's best practice.

tonybach replied:
`<FontContainer>` is a little bit long. Do you think it would be a good idea to name it something like `<Icon value="icon-fist">`? As `icon` is not an HTML element.

ajainarayanan commented on cdap-ui/app/cdap/components/FastAction/SetPreferenceAction/index.js:
```diff
@@ -74,15 +74,16 @@ export default class SetPreferenceAction extends Component {
         <FastActionButton
-          icon={wrenchClasses}
+          icon='icon-wrench'
+          iconClassName={wrenchClasses}
```
shouldn't it be either `className` or `wrenchClasses` be passed on with the `icon` prop? Were there specific reasons for having `icon` and `iconClassName`. My understanding was `icon` accepts a class name that defines an icon.

tonybach replied:
Yeah this was not very clear, will clean it up.

ajainarayanan commented on cdap-ui/app/cdap/components/NamespaceDropdown/index.js:
```diff
@@ -240,7 +241,9 @@ export default class NamespaceDropdown extends Component {
             </div>
-            <span className="fa fa-caret-down float-xs-right" />
+            <span className="float-xs-right">
+              <SVG icon="icon-caret-down" />
```
If we can wrap things inside a span with a class name do we need a `className` prop for `SVG`? We should probably follow one pattern

tonybach replied:
I think it gives more flexibility. Before when we were still using `<i>` font icons, at times we would give the `<i>` a class, while still wrapping it inside another `<span>`.

ajainarayanan reviewed (commented):
Some minor comments.

ajainarayanan commented:
@tonybach Not sure how helpful it would be but we could use something like this https://github.com/kisenka/svg-sprite-loader to minimize #of network requests made.

tonybach commented:
@ajainarayanan Was trying that same library for some time this afternoon without success. Can try again tomorrow as I do think it will help us reduce the # of network requests.

ajainarayanan commented:
Overall code looks good. I am not sure why we changed the prop name from `icon` to `value`. I am not sure what the `value` signifies? Also if the icon is just a css class name then why not just `className` for everything? If the user wants to add an extra class name they can just add it to the `className` property. 

If we still want to have it as a separate prop can we make it a required one? @tonybach Thoughts?

tonybach commented:
@ajainarayanan 
- I changed the prop name from `icon` to `value`, because we changed the component `SVG` to `Icon` (then later `IconSVG`). I thought that writing something like `<IconSVG icon='icon-fist' />` would be too repetitive, so I changed it to `value`. I can change it back to `icon`, or maybe `name`?
- We need the specific icon value, not just for the class name, but also for `path` value in the `<IconSVG>` component, to get the right icon in the SVG sprite. If we just use `className` for everything then we would have to do string parsing to know what icon needs to be rendered. I can make the prop a required one.

ajainarayanan commented:
@tonybach sounds good. `name` being a required prop.

ajainarayanan commented:
👍 🍾 🎉 Looks great! code LGTM. Demo to @bdmogal or Lea and merge!

tonybach commented:
@ajainarayanan Since it's a big change, will push to a cluster for easier testing.

bdmogal commented:
Did some verification. I don't think its possible to completely verify this, but I think one way to test this would be to try what happens on a slow, remote Windows server. Let's merge this PR, and I'll try some known issues on windows servers after an SDK is generated form this branch.

bdmogal commented:
LGTM.
